### DISCUSSION TITLE
Raycaster POC

### DISCUSSION THREAD

**Discussion by grakoczy** [picocomputer/community#132] [Q&A]
Hi,

I've tried to write simple raycaster. It kinda works, but it's really slow. 
Part of this is due to drawing filled rects which you can see when screen is updating.
I know that using floats for calculations is bad idea for 6502.
I've tried to replace some of those operations with integer math with scaling. It's little bit faster but gives me some errors so I've left floating math version for now.
Code is available here: [https://github.com/grakoczy/RP6502-raycasting-llvm-mos](https://github.com/grakoczy/RP6502-raycasting-llvm-mos)
Maybe someone with more skills will be able to make it run faster?
Anyway it's a lot of fun writing programs for this little machine :)

![raycaster](https://github.com/user-attachments/assets/d14ca711-fa07-4ded-ad4f-f4a2bd968dba)

https://github.com/user-attachments/assets/bb923524-42ca-4d9e-8a15-fab1e45e13cd

**Comment by grakoczy** [comment]
Hi,

I've changed calculations from float to fixed point math and now raycasting calculation is few times faster.
Now the biggest problem is slow drawing of filled rectangles.
Is there any way to avoid drawing simple rectangles filled with flat color? Tiles or spritest? 
I've tried to do it with ASCII art in console but this is too slow in 80x30 mode. 
I saw sprite demos and it looks like fill rate is fast enough to fill the screen few times per second, but how many sprites I could have at once? I imagine having 8 pixels wide sprites, scaling them in Y axis to match desired height and then putting them on the screen. Will it work?

**Reply by rumbledethumps** [reply]
The library you're using is ported from a teaching library built for a completely different architecture. The rectangle fill is not anywhere near optimized for the RIA xram. For every pixel drawn, there's a switch on resolution, a multiplication, and a setting of the xram address. You don't need the switch at all. The multiply and xram setting only needs to be done once per horizontal line. You can even replace the multiply with addition for all but the first line of the central part of the walls (the square part). I'm sure you'll find more when you get into it.

**Reply by grakoczy** [reply]
Thanks, I'll try to optimize it for my case. 
One more thing. Now if I want to use 255,255,255 color I have to stick with 16 bit color mode which requires saving two bytes per pixel. Is there any way I could use custom colors but in 8 bit mode?

**Reply by rumbledethumps** [reply]
I can't give an easy answer for which graphics mode to use. But I doubt 16bpp is the best for this.

**Reply by grakoczy** [reply]
As I understand only 16bpp mode doesn't use built in color palette. If I could build custom palette for 8bbp mode then it will be enough for me as I only need couple of colors, couple shades each. ANSI palette is bit too limited.  
I've already done some optimisation for fill_rect and vline and hline drawing and it works much faster! Thanks for the advice.

**Comment by grakoczy** [comment]
Changing from float to fixed point calculations along with optimized rectangle drawing gave quite an speed improvement.
I still have to handle some overflows in fixed point calculations to avoid empty spots.
Here is how it looks now:

https://github.com/user-attachments/assets/94e76002-56c9-49da-99f0-b623b6eb79e3

**Comment by grakoczy** [comment]
Hi,

I've managed to speed this up and added floor shading.
Do do this I've moved from drawing rectangles into filling the buffer while raycasting calculation and then copying the data to vga registers. 
I have to try with texturing the walls next.

https://github.com/user-attachments/assets/8cb875bb-f3df-491b-85ca-d42ca105c9f5

**Reply by rumbledethumps** [reply]
Wow, that's fast. Here's some more ideas...
You have three planes. The map and view can be their own planes with exact dimensions to save memory. The third plane could be a character mode for UI. Then you'll have enough memory for drawing monsters with the sprite system which can do scaling and clip the sides for when behind walls.

**Reply by grakoczy** [reply]
Sounds great, but how can I configure 3 planes with llvm-mos? I've tried to mix bitmap and character mode before but I couldn't get it working. It works fine in cc65 environment but I don't really understand how xregn works, I'm afraid.

**Reply by rumbledethumps** [reply]
xregn is the same as xreg but it needs the parameter count. xreg can only be implemented in assembly, which hasn't been done for llvm. The paint example uses two planes, one for the drawing and one for the palette.

**Comment by grakoczy** [comment]
I've made some progress. Textures seems to work quite well. I'm also using whole 320x180 screen. Still a lot of work but I think I'm going forward. Now I have to add sprites. 

https://github.com/user-attachments/assets/4589368f-e7ce-4ea4-876a-987db0f53ccc

**Reply by rumbledethumps** [reply]
Sorcery. The only mathematics a 6502 can do is addition and logic.

**Comment by grakoczy** [comment]
I wonder if I can use some sprites along with normal bitmap graphics. Now I'm loading converted image as full screen asset like this:

```
rp6502_asset(raycast 0x10000 textures/pixel-320x180.bin)
rp6502_executable(raycast
    pixel-320x180.bin.rp6502
)
```

Video mode is set like this:

 ```
xram0_struct_set(0xFF00, vga_mode3_config_t, x_wrap, true);
  xram0_struct_set(0xFF00, vga_mode3_config_t, y_wrap, true);
  xram0_struct_set(0xFF00, vga_mode3_config_t, x_pos_px, 0);
  xram0_struct_set(0xFF00, vga_mode3_config_t, y_pos_px, 0);
  xram0_struct_set(0xFF00, vga_mode3_config_t, width_px, SCREEN_WIDTH);
  xram0_struct_set(0xFF00, vga_mode3_config_t, height_px, SCREEN_HEIGHT);
  xram0_struct_set(0xFF00, vga_mode3_config_t, xram_data_ptr, 0x0000);
  xram0_struct_set(0xFF00, vga_mode3_config_t, xram_palette_ptr, 0xFFFF);

  xregn(1, 0, 0, 1, 2);
  xregn(1, 0, 1, 3, 3, 3, 0xFF00);
```

Then I draw other stuff as normal pixels, lines etc. It would look much better if instead of simple drawing the compass needle using lines I could put sprite there and just rotate it.  
Can I just set sprite mode on another plane and add some sprites as assets?

**Reply by rumbledethumps** [reply]
You can put sprites on any or all planes until the fill rate is exceeded.

**Comment by grakoczy** [comment]
I've made some improvements to the raycaster code. It faster now so I can render it with higher resolution when moving. 
I was also able to put affinite sprite with the needle and generate custom palette so it looks bit better. 


https://github.com/user-attachments/assets/e02ef4cd-6bef-4496-bfe2-01da3d5a1f55

**Reply by rumbledethumps** [reply]
I posted this to the llvm-mos discord and theirs heads exploded. You should stop by and see if they have questions. https://discord.gg/D9gRm3aznh

**Comment by grakoczy** [comment]
Thanks for the info and link to to the discord 👍

**Comment by grakoczy** [comment]
Sprites time! 
Small performance penalty but it looks quite cool. I had to change textures from 32x32 to 16x16 to fit them with sprites in memory. 

https://github.com/user-attachments/assets/451ae721-63da-424b-b694-425ad4243c25

**Reply by rumbledethumps** [reply]
You're almost to wolf3d. Thew 8 year old child at our last meeting was fully enthralled by this.

**Comment by grakoczy** [comment]
I've added floor/ceiling textures and changed theme. There is also map scan, too. 
You can find all controls on [github](https://github.com/grakoczy/RP6502-raycasting-llvm-mos)
I think I'll stop tinkering with this project for a while. I don't know what to do next. Till now I was just adding new stuff, but I think it's enough for a demo.

https://github.com/user-attachments/assets/16ad6397-2441-40aa-842f-de3b6aaed825
